### PR TITLE
By yanniboi: Added manage mode for managing dashboard blocks.

### PR DESCRIPTION
**Pull request by yanniboi**
This PR provides the manage dashboard toolbar icon that switches the state. It also provides the js and routes to reload the tabs with the correct blocks.

**TEST SITE**: http://contacts.dev.jobboard.freelygive.org.uk/admin/contacts/3

### CODE REVIEW

**Comment by yanniboi** [comment]
selection_267

selection_268

File: contacts.libraries.yml [hunk 1]
```diff
@@ -14,6 +14,7 @@ tabs:
     - core/jquery
     - core/drupal.ajax
     - contacts/drupal.ajax
+    - contacts/dashboard.manage
```
Comment (andrewbelcher) [inline contacts.libraries.yml]: Should this be a dependency? Presumably manage is only for people with manage permissions so should be dealt with separately?

File: contacts.module [hunk 2]
```diff
@@ -40,9 +41,75 @@ function contacts_theme() {
...
+      $variables['label'] = $variables['label'] ?: $variables['block']->label(FALSE);
+      $variables['content']['description'] = [
+        '#type' => 'html_tag',
+        '#tag' => 'strong',
+        '#value' => $variables['description'],
+        '#prefix' => '<div>',
+        '#suffix' => '</div>',
+      ];
+
+      if ($variables['block_plugin'] == 'contacts_entity') {
```
Comment (andrewbelcher) [inline contacts.module]: Could this go into an interface and method on blocks so other blocks can also provide it and we pull block specific code out of here. Or an alternative hook...

File: contacts.module [hunk 3]
```diff
@@ -161,6 +228,52 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+    '#cache' => [
+      'contexts' => [
+        'user.permissions',
+      ],
+    ],
+  ];
+
+  if (!\Drupal::currentUser()->hasPermission('administer contacts')) {
+    return $items;
+  }
```
Comment (andrewbelcher) [inline contacts.module]: Is this 'the right way TM', or do they normally include the items with an `'#access' => FALSE`?

File: contacts.routing.yml [hunk 4]
```diff
@@ -29,9 +29,16 @@ contacts.add_org_form:
   options:
     _admin_route: TRUE
 
-contacts.ajax_update_blocks:
+contacts.ajax.manage_mode:
+  path: '/admin/contacts/ajax/manage-mode'
+  defaults:
+    _controller: '\Drupal\contacts\Controller\DashboardController::ajaxManageMode'
+  requirements:
+    _permission: 'administer contacts'
```
Comment (andrewbelcher) [inline contacts.routing.yml]: Is this the right permission, or do we need a new one for managing tabs etc?

File: css/dashboard.manage.css [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @file
+ * Styling for contacts module dashboard management.
+ */
+
+/* Tab appearance. */
+.toolbar .toolbar-bar .contacts-toolbar-tab.toolbar-tab {
```
Comment (andrewbelcher) [inline css/dashboard.manage.css]: Are we able to reduce the specificity here? May not be able to...

File: js/ajax.js [hunk 6]
```diff
@@ -9,45 +9,35 @@
 
   'use strict';
 
-    // Get our original AJAX behavior so we can still call it.
```
Comment (andrewbelcher) [inline js/ajax.js]: This change may be better to go in separate to this? Well need to up our min version of Drupal along side it.

File: js/tabs.js [hunk 7]
```diff
@@ -25,7 +25,7 @@
     
     if (response.activeTab) {
       $('.contacts-ajax-tabs .' + activeClass).removeClass(activeClass);
-      $('.contacts-ajax-tabs .' + response.activeTab).find('a').andSelf().addClass(activeClass);
+      $('.contacts-ajax-tabs .' + response.activeTab).find('a').addBack().addClass(activeClass);
```
Comment (andrewbelcher) [inline js/tabs.js]: This is presumably also for an update to jquery with a Drupal update? Could go with the ajax.js change?

File: src/ContactsTabManagerInterface.php [hunk 8]
```diff
@@ -69,12 +69,12 @@ public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NUL
...
    *   The tab.
-   * @param \Drupal\user\UserInterface $contact
+   * @param \Drupal\user\UserInterface|null $contact
    *   The contact.
    *
    * @return bool
    *   Whether the tab is valid.
    */
-  public function verifyTab(ContactTabInterface $tab, UserInterface $contact);
+  public function verifyTab(ContactTabInterface $tab, UserInterface $contact = NULL);
```
Comment (andrewbelcher) [inline src/ContactsTabManagerInterface.php]: This is making quite a lot of change, presumably due to not wanting to require validating for a contact for manage mode. Perhaps it's worth considering manage mode having a completely separate pipeline?

File: templates/contacts-dnd-card.html.twig [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    <h5>{{ label }}</h5>
+    {% if content %}
+        <div class="content">
+            {{ content }}
+        </div>
+    {% endif %}
+    {% if footer %}
+        {{ footer }}
+    {% endif %}
+</div>
```
Comment (andrewbelcher) [inline templates/contacts-dnd-card.html.twig]: Missing new line

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
This is looking great! Coming together nicely.

File: js/ajax.js (hunk 6 above)
Comment (yanniboi) [reply js/ajax.js]: Cool, I've already got an MR for that.

File: contacts.module (hunk 3 above)
Comment (yanniboi) [reply contacts.module]: This is how core does it in `contextual_toolbar`:

```
function contextual_toolbar() {
  $items = [];
  $items['contextual'] = [
    '#cache' => [
      'contexts' => [
        'user.permissions',
      ],
    ],
  ];

  if (!\Drupal::currentUser()->hasPermission('access contextual links')) {
    return $items;
  }
```

Comment (andrewbelcher) [reply contacts.module]: Perfect!

File: contacts.libraries.yml [hunk 10]
```diff
@@ -47,6 +47,23 @@ dashboard.layout:
     layout:
       css/dashboard-layout.css: {}
 
+# Contacts dashboard management.
+dashboard.manage:
+  js:
+    # Ensure to run before contextual/drupal.context-toolbar.
+    # Core.
```
Comment (andrewbelcher) [inline contacts.libraries.yml]: This comment looks odd?

File: contacts.module [hunk 11]
```diff
@@ -40,9 +42,63 @@ function contacts_theme() {
...
+ */
+function template_preprocess_contacts_dnd_card(array &$variables) {
+  $configuration = $variables['block']->getConfiguration();
+
+  list($variables['block_plugin'], $variables['identifier']) = explode(PluginBase::DERIVATIVE_SEPARATOR, $variables['id'], 2);
+
+  $variables['attributes']['class'][] = 'dnd-block';
+  $variables['attributes']['data-dnd-contacts-block-id'] = $variables['id'];
+  $variables['attributes']['data-dnd-contacts-block-name'] = $configuration['name'];
+  $variables['attributes']['data-dnd-contacts-block-user'] = $variables['user'];
```
Comment (andrewbelcher) [inline contacts.module]: Do the D-n-d cards have a user?

File: contacts.module [hunk 12]
```diff
@@ -40,9 +42,63 @@ function contacts_theme() {
...
+      $variables['content']['description'] = [
+        '#type' => 'html_tag',
+        '#tag' => 'strong',
+        '#value' => $variables['description'],
+        '#prefix' => '<div>',
+        '#suffix' => '</div>',
+      ];
+
+      if ($variables['block'] instanceof DashboardBlockInterface) {
+        $variables['block']->processDragMode($variables);
```
Comment (andrewbelcher) [inline contacts.module]: Is 'drag mode' the best description? Perhaps `processManageMode`?

File: contacts.module [hunk 13]
```diff
@@ -161,6 +217,51 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+      ],
+    ],
+    'tray' => [],
+    '#wrapper_attributes' => [
+      'class' => [
+        'contacts-toolbar-tab',
+        'visually-hidden',
+      ],
+    ],
+  ];
```
Comment (andrewbelcher) [inline contacts.module]: Should we be adding in any libraries to manage it here?

File: contacts.module [hunk 14]
```diff
@@ -161,6 +217,51 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+  $items = [];
+  $items['manage_mode'] = [
+    '#cache' => [
+      'contexts' => [
+        'user.permissions',
+      ],
+    ],
+  ];
+
+  if (!\Drupal::currentUser()->hasPermission('manage contacts dashboard')) {
```
Comment (andrewbelcher) [inline contacts.module]: Does this also need to be only on the dashboard pages?

File: src/ContactsTabManager.php [hunk 15]
```diff
@@ -156,7 +141,7 @@ public function getTabs(UserInterface $contact) {
   /**
    * {@inheritdoc}
    */
-  public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NULL, $verify = TRUE) {
+  public function getBlocks(ContactTabInterface $tab, $apply_context = FALSE, UserInterface $contact = NULL) {
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php]: Probably can just apply context if `$contact !== NULL`? Is there ever a case where we want to apply context but may not have a user?

File: src/ContactsTabManagerInterface.php [hunk 16]
```diff
@@ -13,64 +13,57 @@
...
    *   The path for the tab.
    *
    * @return \Drupal\contacts\Entity\ContactTab|null
    *   The tab, if it exists.
    */
-  public function getTabByPath(UserInterface $contact, $path);
+  public function getTabByPath($path);
 
   /**
    * Get the tabs for a contact.
```
Comment (andrewbelcher) [inline src/ContactsTabManagerInterface.php]: No longer for a contact.

Comment (andrewbelcher) [reply src/ContactsTabManagerInterface.php]: I wonder if it'd still be useful being able to get tabs for a contact which encapsulates all the logic? We could make the contact optional and, if provided, handle the verification internally?

File: src/ContactsTabManagerInterface.php [hunk 17]
```diff
@@ -13,64 +13,57 @@
...
+  public function getBlocks(ContactTabInterface $tab, $apply_context = FALSE, UserInterface $contact = NULL);
 
   /**
    * Verify that a tab is valid for the given contact.
    *
    * @param \Drupal\contacts\Entity\ContactTabInterface $tab
    *   The tab.
-   * @param \Drupal\user\UserInterface $contact
-   *   The contact.
+   * @param \Drupal\user\UserInterface| $contact
```
Comment (andrewbelcher) [inline src/ContactsTabManagerInterface.php]: Shouldn't have the `|`

File: src/Controller/DashboardController.php [hunk 18]
```diff
@@ -96,11 +117,44 @@ public function ajaxTab(UserInterface $user, $subpage) {
     $response = new AjaxResponse();
     $response->addCommand(new ContactsTab($subpage, $url->toString()));
     $response->addCommand(new HtmlCommand('#contacts-tabs-content', $content));
+    $response->addCommand(new HtmlCommand('#contacts-tabs', $tabs));
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: Feel like these should be the other way round :P

File: src/Controller/DashboardController.php [hunk 19]
```diff
@@ -96,11 +117,44 @@ public function ajaxTab(UserInterface $user, $subpage) {
...
+    if (is_null($manage_mode)) {
+      // Toggle manage mode.
+      $manage_mode = $this->state()->get('manage_mode');
+    }
+
+    $referer = \Drupal::request()->server->get('HTTP_REFERER');
+    $fake_request = Request::create($referer);
+
+    /* @var \Drupal\Core\Url $url_object */
+    $url_object = \Drupal::service('path.validator')->getUrlIfValid($fake_request->getRequestUri());
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php]: Can we inject all this stuff please?

File: src/ContactsTabManager.php [hunk 20]
```diff
@@ -156,7 +141,7 @@ public function getTabs(UserInterface $contact) {
...
     $blocks = $tab->getBlockPlugins();
     if (empty($blocks)) {
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php]: I think we need to differentiate between empty (i.e. no valid blocks) and not attempted... Otherwise a call to `getTab()` followed by `getBlocks()` could result in the whole process happening twice to return empty.

File: src/Element/ContactTabContent.php [hunk 21]
```diff
@@ -68,8 +70,14 @@ public function getInfo() {
    */
   public static function preRenderTabContent(array $element) {
     $tab_manager = static::getTabManager();
-    // Check this tab is valid for the contact.
-    if ($element['#tab'] && $tab_manager->verifyTab($element['#tab'], $element['#user'])) {
+
+    // Verify tab if necesary.
+    if (!$element['#manage_mode']) {
+      $tab_manager->verifyTab($element['#tab'], $element['#user']);
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: If the tab fails to verify at all, we should presumable return here?

File: src/Element/ContactTabContent.php [hunk 22]
```diff
@@ -68,8 +70,14 @@ public function getInfo() {
...
+    }
+
+    $blocks = $tab_manager->getBlocks($element['#tab'], !$element['#manage_mode'], $element['#user']);
+    if (!empty($blocks)) {
       $layout = $element['#tab']->get('layout') ?: 'contacts_tab_content.stacked';
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: Is stacked our default rather than a single region layout?

File: src/Element/ContactTabContent.php [hunk 23]
```diff
@@ -80,28 +88,44 @@ public static function preRenderTabContent(array $element) {
...
-          '#weight' => $block->getConfiguration()['weight'],
-          'content' => $block->build(),
-        ];
-
-        // Add edit link to title.
-        if ($block instanceof DashboardBlockInterface) {
-          $block_content['#dashboard_label_edit_link'] = $block->getEditLink(DashboardBlockInterface::EDIT_LINK_TITLE);
-          $block_content['#pre_render'][] = 'contacts_dashboard_block_edit_link_pre_render';
+        // For some reason build() brings in the theme hooks required...
+        $content = $block->build();
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: Do we need this for manage mode? If not - it can go inside the else.

File: src/Element/ContactTabContent.php [hunk 24]
```diff
@@ -80,28 +88,44 @@ public static function preRenderTabContent(array $element) {
...
+            ],
+            '#id' => $block->getPluginId(),
+            '#block' => $block,
+            '#user' => $element['#user']->id(),
+            '#subpage' => $element['#subpage'],
+            '#mode' => 'manage',
+          ];
+        }
+        else {
+          // @todo fix weight.
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: What is meant by this?

File: src/Element/ContactTabs.php [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+    return $element;
+  }
+
+  /**
+   * Gets the tab manager service.
+   *
+   * @return \Drupal\contacts\ContactsTabManager
+   *   The tab manager service.
+   */
+  protected static function getTabManager() {
```
Comment (andrewbelcher) [inline src/Element/ContactTabs.php]: These feel like they could go into a trait?

File: src/Plugin/Block/ContactsEntity.php [hunk 26]
```diff
@@ -163,6 +187,11 @@ public function getEditLink($mode) {
     }
 
     $params = $this->routeMatch->getRawParameters()->all();
+
+    if (empty($params['user']) || empty($params['subpage'])) {
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: Why is this needed? If we don't have that, we shouldn't be rendering this block, but the DnD for it..?

File: src/Plugin/Block/ContactsEntity.php [hunk 27]
```diff
@@ -178,6 +207,24 @@ public function getEditLink($mode) {
     return $link;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function processDragMode(array &$variables) {
+    $definition = $this->blockManager->getDefinition($variables['id']);
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: Is this not available from `$this->definition`?

File: src/Plugin/Block/ContactsEntity.php [hunk 28]
```diff
@@ -380,4 +427,55 @@ public function calculateDependencies() {
     return $dependencies;
   }
 
+  /**
+   * Get list of links to display on manage block.
+   *
+   * @return array
+   *   Array of links to be based to an 'item_list' render array.
+   */
+  public function getManageLinks() {
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: Should this be protected?

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
These changes are great! Few minor tidy up/questions.

The main question is why you've moved to using a render element for `ContactTabs` rather than letting the block handle it... As render elements handle everything statically, they should generally only be used where everything is being passed in...

File: contacts.module [hunk 29]
```diff
@@ -161,6 +214,58 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
...
+/**
+ * Implements hook_toolbar().
+ */
+function contacts_toolbar() {
+  $items = [];
+  $items['manage_mode'] = [
+    '#cache' => [
+      'contexts' => [
+        'user.permissions',
+        'url.path',
```
Comment (andrewbelcher) [inline contacts.module]: As all our stuff is based on one route, can we use route? Means it will cache across all URLs for that route..?

File: contacts.module [hunk 30]
```diff
@@ -161,6 +214,58 @@ function template_preprocess_contacts_dash_summary(array &$variables) {
   }
 }
 
+/**
+ * Implements hook_toolbar().
+ */
+function contacts_toolbar() {
+  $items = [];
+  $items['manage_mode'] = [
```
Comment (andrewbelcher) [inline contacts.module]: Is this name too prone to collisions? `contacts_manage`?

File: contacts.permissions.yml [hunk 31]
```diff
@@ -3,6 +3,10 @@ administer contacts:
   description: 'Configure the contacts dashboard and related settings.'
   restrict access: TRUE
 
+manage contacts dashboard:
+  title: 'Manage contacts dashboard'
+  description: 'Add, remove and rearrange dashboard tab content.'
```
Comment (andrewbelcher) [inline contacts.permissions.yml]: Could probably just be 'Configure dashboard tabs and tab content', as we'll also be able to configure them?

File: images/pencil-active.svg [hunk 32]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><g><path fill="#ffffff" d="M14.545 3.042l-1.586-1.585c-.389-.389-1.025-.389-1.414 0l-1.293 1.293 3 3 1.293-1.293c.389-.389.389-1.026 0-1.415z"/><rect fill="#ffffff" x="5.129" y="3.8" transform="matrix(-.707 -.707 .707 -.707 6.189 20.064)" width="4.243" height="9.899"/><path fill="#ffffff" d="M.908 14.775c-.087.262.055.397.316.312l2.001-.667-1.65-1.646-.667 2.001z"/></g></svg>
```
Comment (andrewbelcher) [inline images/pencil-active.svg]: If the only difference is the fill colour, can we just use [`.active{ fill: #fff; }`](https://css-tricks.com/almanac/properties/f/fill/)?

File: css/dashboard.manage.css [hunk 33]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @file
+ * Styling for contacts module dashboard management.
+ */
+
+/* Tab appearance. */
```
Comment (andrewbelcher) [inline css/dashboard.manage.css]: Should `contacts.icons.css` and `dashboard.manage.css` be combined into `contacts.toolbar.css`?

File: src/ContactsTabManager.php [hunk 34]
```diff
@@ -156,7 +150,7 @@ public function getTabs(UserInterface $contact) {
   /**
    * {@inheritdoc}
    */
-  public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NULL, $verify = TRUE) {
+  public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NULL) {
     $blocks = $tab->getBlockPlugins();
     if (empty($blocks)) {
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php]: Think we still want this to be an `if (!isset($blocks))` with the default being NULL, so we can tell the difference between no blocks and blocks not gathered?

File: src/Controller/DashboardController.php [hunk 35]
```diff
@@ -95,12 +154,45 @@ public function ajaxTab(UserInterface $user, $subpage) {
...
+      // Toggle manage mode.
+      $manage_mode = $this->state()->get('manage_mode');
+    }
+
+    // Get parameters from referrer request.
+    $referrer = $this->request->server->get('HTTP_REFERER');
+    $fake_request = Request::create($referrer);
+
+    /* @var \Drupal\Core\Url $url_object */
+    $url_object = $this->pathValidator->getUrlIfValid($fake_request->getRequestUri());
```
Comment (andrewbelcher) [inline src/Controller/DashboardController.php:183]: Here's a wacky idea... what about a redirect to the referring page?

File: src/Element/ContactTabContent.php [hunk 36]
```diff
@@ -67,9 +63,8 @@ public function getInfo() {
...
    */
   public static function preRenderTabContent(array $element) {
-    $tab_manager = static::getTabManager();
-    // Check this tab is valid for the contact.
-    if ($element['#tab'] && $tab_manager->verifyTab($element['#tab'], $element['#user'])) {
+    $blocks = $element['#blocks'];
+    if (!empty($blocks)) {
       $layout = $element['#tab']->get('layout') ?: 'contacts_tab_content.stacked';
       $layout_manager = static::getLayoutManager();
       $layoutInstance = $layout_manager->createInstance($layout, []);
```
Comment (andrewbelcher) [inline src/Element/ContactTabContent.php]: I suppose we could pass the layout instance in in the same way we do the blocks, and ditch using services internally completely?

Comment (andrewbelcher) [reply src/Element/ContactTabContent.php]: Loking at the code below, we only use it to get the regions? In which case - let's just pass in the regions!

Comment (andrewbelcher) [reply src/Element/ContactTabContent.php]: Or do we even need to do that as the blocks know what region they're in, do we need empty regions to exist?

File: src/Plugin/Block/ContactsDashboardTabs.php [hunk 37]
```diff
@@ -100,49 +126,29 @@ public function build() {
    *   Drupal renderable array being added to.
    */
   public function buildTabs(array &$build) {
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsDashboardTabs.php]: Should this be `protected`?

File: src/Plugin/Block/ContactsDashboardTabs.php [hunk 38]
```diff
@@ -152,13 +158,25 @@ public function buildTabs(array &$build) {
    *   Drupal renderable array being added to.
    */
   public function buildContent(array &$build) {
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsDashboardTabs.php]: Should this be `protected`?

File: src/Plugin/Block/ContactsDashboardTabs.php [hunk 39]
```diff
@@ -27,6 +28,20 @@ class ContactsDashboardTabs extends BlockBase implements ContextAwarePluginInter
    */
   protected $tabManager;
 
+  /**
+   * The state service.
+   *
+   * @var \Drupal\Core\State\StateInterface
+   */
+  protected $stateService;
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsDashboardTabs.php]: Just `$state` is nicer :)

File: src/Plugin/Block/ContactsEntity.php [hunk 40]
```diff
@@ -380,4 +419,55 @@ public function calculateDependencies() {
...
+  protected function getManageLinks() {
+    $entity_id = $this->pluginDefinition['_entity_type_id'];
+    $bundle_id = $this->pluginDefinition['_bundle_id'];
+    $entity_definition = $this->entityTypeManager->getDefinition($entity_id);
+    $bundle_type = $entity_definition->getBundleEntityType();
+    $operations = [];
+    // Add manage fields and display links if this entity type is the bundle
+    // of another and that type has field UI enabled.
+    if ($bundle_type && $entity_definition->get('field_ui_base_route')) {
+      $account = $this->currentUser;
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: No need to assign to a variable as it's not a getter, just use `$this->currentUser->hasPermission()`

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
Oooo it feels soooo close!

File: src/Controller/DashboardController.php (hunk 35 above)
Comment (yanniboi) [reply src/Controller/DashboardController.php:183]: As in, instead of ajax?

File: css/contacts.toolbar.css [hunk 41]
```diff
@@ -0,0 +1,36 @@
...
+}
+.toolbar-bar .toolbar-icon-page-manager-page-view-contacts-dashboard:active:before,
+.toolbar-bar .toolbar-icon-page-manager-page-view-contacts-dashboard.active:before {
+  background-image: url(../images/contacts.png);
+}
+
+.toolbar-bar .toolbar-icon-edit:before {
+  background-color: #bebebe;
+  mask: url(../images/pencil.svg) no-repeat 50% 50%;
+  -webkit-mask: url(../images/pencil.svg) no-repeat 50% 50%;
```
Comment (andrewbelcher) [inline css/contacts.toolbar.css:20]: Masks?!

File: src/ContactsTabManager.php [hunk 42]
```diff
@@ -166,67 +160,78 @@ public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NUL
...
+          $this->contextHandler->applyContextMapping($block, $contexts);
+        }
+        catch (ContextException $exception) {
+          unset($blocks[$key]);
+          continue;
+        }
       }
+
+      // @todo Don't allow block to be added if relationships have not been
+      // satisfied.
```
Comment (andrewbelcher) [inline src/ContactsTabManager.php]: I think this is already done with the code above? In which case this can be removed...

File: src/Controller/DashboardController.php (hunk 35 above)
Comment (andrewbelcher) [reply src/Controller/DashboardController.php:183]: Let's ignore this for now!

File: src/Plugin/Block/ContactsEntity.php [hunk 43]
```diff
@@ -178,6 +198,24 @@ public function getEditLink($mode) {
     return $link;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function processManageMode(array &$variables) {
+    $definition = $this->blockManager->getDefinition($variables['id']);
```
Comment (andrewbelcher) [inline src/Plugin/Block/ContactsEntity.php]: Can this not just use `$this->definition` and avoid injecting block manager entirely?

Comment (yanniboi) [reply src/Plugin/Block/ContactsEntity.php]: doh!

File: css/contacts.toolbar.css (hunk 41 above)
Comment (yanniboi) [reply css/contacts.toolbar.css:20]: Yup. `fill` doesnt work if the SVG is a `background-image`. But this does the trick.

**Comment by andrewbelcher** [comment]
I'm happy! Merge at your leisure!
